01_download-data/physprop/final/filter-data-training.py

```diff
@@ -0,0 +1,294 @@
+    """
+    Save the dataset to a CSV and JSON file.
+    The CSV file will be a valid PhysicalPropertyDataSet CSV file.
+    The JSON file will be a valid PhysicalPropertyDataSet JSON file.
+    """
+    output_file.parent.mkdir(parents=True, exist_ok=True)
+    dataset.to_pandas().to_csv(output_file)
+    dataset.json(output_file.with_suffix(".json"), format=True)
+
+    logger.info(f"Saved to {output_file}")
+    logger.info(f"Saved to {output_file.with_suffix('.json')}")
+
+
+@click.command()
+@click.option(
+    "--output-file",
+    "-o",
+    default="output/training-set.csv",
+    help=(
+        "The output CSV file to save the filtered data to. "
+        "Note, a JSON file with the same name but with a .json extension will also be created. "
+        "Both encode PhysicalPropertyDataSet objects."
+    ),
+)
+@click.option(
+    "--input-file",
+    "-i",
+    default="../intermediate/output/renamed-filtered.csv",
+    help="The CSV file containing existing parsed ThermoML data",
+)
+@click.option(
+    "--exclude-file",
+    "-x",
+    default=None,
+    help="The file containing SMILES to exclude",
+)
+@click.option(
+    "--n-processes",
+    "-np",
+    default=1,
+    help="The number of processes to use for filtering the data",
+)
+def main(
+    input_file: str = "../intermediate/output/renamed-filtered.csv",
+    exclude_file: str = None,
+    output_file: str = "output/training-set.csv",
+    n_processes: int = 1,
+):
+    now = time.time()
+    logger.info(f"Starting at {time.ctime(now)}")
+
+    df = pd.read_csv(input_file)
+    df["Id"] = df["Id"].astype(str)
+    df["N Components"] = df["N Components"].astype(int)
+    logger.info(f"Loaded dataframe {time.ctime(now)}")
+
+    ds = PhysicalPropertyDataSet.from_pandas(df)
+    logger.info(f"Converted dataframe to PhysicalPropertyDataSet at {time.ctime(now)}")
+
+    thermoml_data_frame = ds.to_pandas()
+    logger.info(f"Loading {len(thermoml_data_frame)} data")
+
+    # load smiles to exclude
+    if exclude_file:
+        with open(exclude_file, "r") as f:
+            contents = f.readlines()
+        smiles_to_exclude = [x.strip().split()[0] for x in contents]
+    else:
+        smiles_to_exclude = []
+
+    property_type_filter = filtering.FilterByPropertyTypesSchema(
+        property_types=[
+            "Density",
+            "EnthalpyOfMixing",
+        ],
+        n_components={
+            "Density": [1, 2],
+            "EnthalpyOfMixing": [2],
+        },
+        strict=True,
+    )
+
+    training_set_frame = curate_data_set(
+        thermoml_data_frame,
+        smiles_to_exclude,
+        property_type_filter,
+        n_processes,
+    )
+    logger.info(f"Filtered to {len(training_set_frame)} data points")
+
+    assert len(training_set_frame) > 0, "No data points left after filtering"
+    # make sure we wind up with a reasonable number of data points
+    assert len(training_set_frame) > 1000, "Not enough data points left after filtering"
```

Comment:
The execution here LGTM, but I'd be concerned about the dataset size -- 1000 properties will take a while to run. Is this the final training dataset you plan to run or do you plan to add more filters? If so the one we discussed this morning (where at least one molecule includes a N) seems it'd be a good start.

Reply:
I'd suggest aiming for ~100 properties for training just for summer camp.